spmd/tensor/api.py

```diff
@@ -30,41 +86,53 @@ class Tensor(torch.Tensor):  # pyre-ignore[13]: pyre is bad at __new__
     @staticmethod
     def __new__(
         cls,
+        local_tensor: torch.Tensor,
         device_mesh: DeviceMesh,
         placements: Sequence[Placement],
-        size: torch.Size,
         # pyre-fixme[2]: Parameter must be annotated.
         **kwargs,
     ) -> "Tensor":
-        # new method instruct wrapper tensor and add placement spec
-        # it does not do actual distribution, __init__ should do it instead.
-        # TODO: implement __init__ for tensor constructors
-        assert isinstance(placements, list)
-        # sizes = _flatten_tensor_size(size)
-        dtype = kwargs["dtype"]
-        layout = kwargs["layout"]
-        requires_grad = kwargs["requires_grad"]
-        device = kwargs.get("device", "cpu")
+        tensor_shape = list(local_tensor.size())
+        # recover tensor shape in the case of sharding
+        # TODO: also recover the strides in the case of sharding
+        for idx, placement in enumerate(placements):
+            if isinstance(placement, Shard):
+                shard_dim = placement.dim
+                # recover tensor shape on the shard dim
+                tensor_shape[shard_dim] = tensor_shape[
+                    shard_dim
+                ] * device_mesh.size(idx)
+            elif not isinstance(placement, (Replicate, _Partial)):
+                raise RuntimeError(
+                    f"placement type {type(placement)} not supported!"
+                )
 
+        # new method instruct wrapper tensor from local_tensor and add
+        # placement spec, it does not do actual distribution
         r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
             cls,
-            size,
-            dtype=dtype,
-            device=device,
-            layout=layout,
-            requires_grad=requires_grad,
+            torch.Size(tensor_shape),
+            strides=local_tensor.stride(),
+            dtype=local_tensor.dtype,
+            device=local_tensor.device,
+            layout=local_tensor.layout,
+            requires_grad=local_tensor.requires_grad,
```

Comment:
This should be `kwargs.get("requires_grad")`
```python
>>> a = torch.rand(2, requires_grad=True)
>>> torch.tensor(a)
<stdin>:1: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
tensor([0.1073, 0.9690])
>>> torch.tensor(a, requires_grad=True)
tensor([0.1073, 0.9690], requires_grad=True)
>>> torch.tensor(a.detach(), requires_grad=True)
tensor([0.1073, 0.9690], requires_grad=True)

```